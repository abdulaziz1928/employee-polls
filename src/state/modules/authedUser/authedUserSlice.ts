import { PayloadAction, createSlice } from "@reduxjs/toolkit";

const initialState: string | null = null;

const authedUserSlice = createSlice({
  name: "authedUser",
  initialState: initialState as string | null,
  reducers: {
    setAuthedUser: (_, action: PayloadAction<string>) => action.payload,
    logoutAuthedUser: () => null,
  },
});

export const { setAuthedUser, logoutAuthedUser } = authedUserSlice.actions;

export default authedUserSlice.reducer;
