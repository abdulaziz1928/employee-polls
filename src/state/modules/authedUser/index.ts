import reducer from "./authedUserSlice";
export * from "./authedUserSlice";
export default reducer;
