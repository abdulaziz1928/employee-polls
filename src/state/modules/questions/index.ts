import reducer from "./questionsSlice";
export * from "./questionsSlice";
export * from "./models/question";
export default reducer;
