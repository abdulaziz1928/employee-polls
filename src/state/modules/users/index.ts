import reducer from "./usersSlice";
export * from "./usersSlice";
export * from "./models/user";
export default reducer;
