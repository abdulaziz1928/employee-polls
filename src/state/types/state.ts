import LoadingStatus from "./loading_status";

export default interface IState<T> {
  entities: T;
  loading: LoadingStatus;
  error?: string;
}
