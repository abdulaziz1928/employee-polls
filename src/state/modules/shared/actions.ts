import { Dispatch } from "@reduxjs/toolkit";
import { setAuthedUser } from "../authedUser";
import { getInitialData } from "../../apis/api";
import { receiveQuestions } from "../questions";
import { receiveUsers } from "../users";

const AUTHED_USER_ID = "tylermcginnis";

export function handleInitialData() {
  return (dispatch: Dispatch) => {
    return getInitialData().then(({ users, questions }) => {
      dispatch(receiveUsers(users));
      dispatch(receiveQuestions(questions));
      dispatch(setAuthedUser(AUTHED_USER_ID));
    });
  };
}
