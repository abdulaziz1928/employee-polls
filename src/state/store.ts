import authedUser from "./modules/authedUser";
import questions from "./modules/questions";
import users from "./modules/users";
import logger from "./middlewares/logger";
import prefrences from "./modules/prefrences/prefrencesSlice";
import snackbar from "./modules/snackbar/snackbarSlice";
import {
  PreloadedState,
  combineReducers,
  configureStore,
} from "@reduxjs/toolkit";

const reducers = { authedUser, users, questions, prefrences, snackbar };

const rootReducer = combineReducers(reducers);

const middleWares = [logger];

export const setupStore = (preloadedState?: PreloadedState<RootState>) => {
  return configureStore({
    reducer: rootReducer,
    middleware: (gDM) => gDM().concat(...middleWares),
    preloadedState,
  });
};

export type RootState = ReturnType<typeof rootReducer>;
export type AppStore = ReturnType<typeof setupStore>;
export type AppDispatch = AppStore["dispatch"];
