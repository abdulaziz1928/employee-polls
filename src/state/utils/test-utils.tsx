import React, { PropsWithChildren } from "react";
import { render } from "@testing-library/react";
import type { RenderOptions } from "@testing-library/react";
import type { PreloadedState } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import { setupStore, AppStore, RootState } from "../store";
import { MemoryRouter } from "react-router-dom";

interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export function renderWithProvidersAndRouter(
  ui: React.ReactElement,

  {
    preloadedState = {},
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {},
  route: string = "/"
) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return (
      <Provider store={store}>
        <MemoryRouter initialEntries={[route]}>{children}</MemoryRouter>
      </Provider>
    );
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
