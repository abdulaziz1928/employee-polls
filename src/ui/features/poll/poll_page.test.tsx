import { Route, Routes } from "react-router-dom";
import { setAuthedUser } from "../../../state/modules/authedUser";
import { AppStore, setupStore } from "../../../state/store";
import { renderWithProvidersAndRouter } from "../../../state/utils/test-utils";
import PollPage from "./poll_page";
import { handleInitialData } from "../../../state/modules/shared/actions";
import { fireEvent, screen } from "@testing-library/react";
import PageRoutes from "../../../state/types/page_routes";

let store: AppStore;
const page = (
  <Routes>
    <Route path={PageRoutes.Question} element={<PollPage />} />
  </Routes>
);
describe("PollPage", () => {
  beforeAll(async () => {
    store = setupStore();
    await store.dispatch(handleInitialData());
    store.dispatch(setAuthedUser("tylermcginnis"));
  });

  it("should render the component", () => {
    const { container } = renderWithProvidersAndRouter(<PollPage />);
    expect(container).toBeDefined();
    expect(container).toMatchSnapshot();
  });

  it("should render AnswerPoll form if user has not voted on the poll", () => {
    renderWithProvidersAndRouter(
      page,
      { store },
      "/questions/8xf0y6ziyjabvozdd253nd"
    );
    expect(screen.getByTestId("answer-poll-container")).toBeInTheDocument();
  });

  it("should render PollResults if user has already voted on the poll", () => {
    renderWithProvidersAndRouter(
      page,
      { store },
      "/questions/vthrdm985a262al8qx3do"
    );
    expect(screen.getByTestId("poll-results-container")).toBeInTheDocument();
  });

  it("should redirect to PollResults once the user votes on an option", async () => {
    renderWithProvidersAndRouter(
      page,
      { store },
      "/questions/8xf0y6ziyjabvozdd253nd"
    );
    expect(screen.getByTestId("answer-poll-container")).toBeInTheDocument();

    const submitButton = screen.getByTestId("submit-choice-one-container");

    expect(submitButton).toBeInTheDocument();

    fireEvent.click(submitButton);
    await screen.findByTestId("poll-results-container");

    expect(screen.getByTestId("poll-results-container")).toBeInTheDocument();
    expect(
      screen.queryByTestId("answer-poll-container")
    ).not.toBeInTheDocument();
  });
});
