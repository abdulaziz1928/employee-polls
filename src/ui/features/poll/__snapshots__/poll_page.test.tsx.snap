// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PollPage should render the component 1`] = `
<div>
  <div
    class="MuiBox-root css-8atqhb"
  >
    <span
      class="MuiLinearProgress-root MuiLinearProgress-colorSecondary MuiLinearProgress-indeterminate css-10jyip1-MuiLinearProgress-root"
      role="progressbar"
    >
      <span
        class="MuiLinearProgress-bar MuiLinearProgress-barColorSecondary MuiLinearProgress-bar1Indeterminate css-p53mz9-MuiLinearProgress-bar1"
      />
      <span
        class="MuiLinearProgress-bar MuiLinearProgress-barColorSecondary MuiLinearProgress-bar2Indeterminate css-xybj4j-MuiLinearProgress-bar2"
      />
    </span>
  </div>
</div>
`;
